inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/feature/FactLinkingServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import javax.annotation.Resource;
+
+import org.apache.uima.cas.CASException;
+import org.apache.uima.cas.text.AnnotationFS;
+import org.apache.uima.jcas.JCas;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import de.tudarmstadt.ukp.clarin.webanno.api.AnnotationSchemaService;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.action.AnnotationActionHandler;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.model.AnnotatorState;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.model.LinkWithRoleModel;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.util.WebAnnoCasUtil;
+import de.tudarmstadt.ukp.clarin.webanno.model.AnnotationFeature;
+import de.tudarmstadt.ukp.clarin.webanno.model.AnnotationLayer;
+import de.tudarmstadt.ukp.clarin.webanno.model.Project;
+import de.tudarmstadt.ukp.inception.kb.KnowledgeBaseService;
+import de.tudarmstadt.ukp.inception.kb.graph.KBHandle;
+import de.tudarmstadt.ukp.inception.kb.graph.KBStatement;
+import de.tudarmstadt.ukp.inception.kb.model.KnowledgeBase;
+
+@Component(FactLinkingService.SERVICE_NAME)
+public class FactLinkingServiceImpl implements FactLinkingService
+{
+    @Resource private KnowledgeBaseService kbService;
+    @Resource private AnnotationSchemaService annotationService;
+
+    private static final String FACT_LAYER = "webanno.custom.Fact";
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public KnowledgeBase getKBByKBHandle(KBHandle kbHandle, Project aProject)
+    {
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(aProject)) {
+            if (kbService.listProperties(kb, false).contains(kbHandle)) {
+                return kb;
+            }
+            if (kbService.listConcepts(kb, false).contains(kbHandle)) {
+                return kb;
+            }
+            for (KBHandle concept : kbService.listConcepts(kb, false)) {
+                if (kbService.listInstances(kb, concept.getIdentifier(), false)
+                    .contains(kbHandle)) {
+                    return kb;
+                }
+            }
+
+        }
+        return null;
+    }
+
+    @Override
+    public List<KBHandle> getKBConceptsAndInstances(Project aProject)
+    {
+        List<KBHandle> handles = new LinkedList<>();
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(aProject)) {
+            handles.addAll(kbService.listConcepts(kb, false));
+            for (KBHandle concept : kbService.listConcepts(kb, false)) {
+                handles.addAll(kbService.listInstances(kb, concept.getIdentifier(), false));
+            }
+        }
+        return new ArrayList<>(handles);
+    }
+
+    @Override
+    public List<KBHandle> getAllPredicatesFromKB(Project aProject)
+    {
+        List<KBHandle> handles = new ArrayList<>();
```

Comment:
I would write Sparql for that. Also, it will break for larger, e.g. remote knowledge bases.

Reply:
I think actually, when user creates a new fact layer, a Knowledgebase should be chosen. So we just need to list the predicates of that layer. WDYT @reckart ?

Reply:
Good question. I think I would agree. If we allow that facts are linkable across KBs, that could make things quite complicated.

Reply:
have created a new issue for it.